### ISSUE TITLE
Crash with CrossValidationConfig and UserMinibatchSource

### ISSUE THREAD

**Issue by bencherian** [Microsoft/CNTK#1894]
I am using objects from a child class of `UserMinibatchSource` as minibatch sources for a `TrainingSession` in Python. Specifically, I am using them for both training data and as an argument to the constructor of a `CrossValidationConfig` object. However, whenever I try to actually use the `CrossValidationConfig` object in the `TrainingSession` the training script crashes on the first call to the cross validation routine with a somewhat bizarre error:

```Python
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
RecursionError: maximum recursion depth exceeded

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
C:\Anaconda3\envs\cntk-py36\lib\site-packages\cntk\cntk_py.py in get_checkpoint_state(self)
   2415 
   2416     def get_checkpoint_state(self):
-> 2417         return _cntk_py.MinibatchSource_get_checkpoint_state(self)
   2418 
   2419     def restore_from_checkpoint(self, arg2):

RuntimeError: SWIG director method error.

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
C:\Anaconda3\envs\cntk-py36\lib\site-packages\cntk\cntk_py.py in get_checkpoint_state(self)
   2415 
   2416     def get_checkpoint_state(self):
-> 2417         return _cntk_py.MinibatchSource_get_checkpoint_state(self)
   2418 
   2419     def restore_from_checkpoint(self, arg2):

RuntimeError: SWIG director method error.
```
(The runtime error in `get_checkpoint_state` is repeated many times...)

I'm completely unfamiliar with SWIG, but I believe this is due to the C++ cross validation code attempting to save and restore the cross validation MinibatchSource's checkpoint state around actually running a test using data from it. 

From Source/CNTKv2LibraryDLL/TrainingSession.cpp:
```cpp
bool TrainingSession::CrossValidate(size_t currentIndex, const DeviceDescriptor& computeDevice)
{
    ...
    auto checkpoint = m_cv.m_source->GetCheckpointState(); //Why do we need to save state?
    bool shouldCV = true;
    while (shouldCV)
    {
        GetCrossValidationMinibatch(minibatch, m_cv.m_mbSize[totalNumberOfSamples], computeDevice);

        // TODO: it may be slow to rely on TestMinibatch to return error each time, since it may require transfer
        // of error from the GPU each time, accumulatedError can be allocated on GPU
        shouldCV = m_trainer->TestMinibatch(minibatch, errorAndCount, computeDevice, m_numberOfWorkers != 1);
        if (shouldCV)
        {
            accumulatedError += errorAndCount.first->AsScalar<double>();
            totalNumberOfSamples += errorAndCount.second;
            numberOfMinibatches++;
        }
    }

    m_cv.m_source->RestoreFromCheckpoint(checkpoint); //Why do we need to restore MinibatchSource state after the test?
       ...
}
```

UserMinibatchSources cannot save a checkpoint state and I'm guessing  the SWIG code gets stuck in some sort of infinite recursion loop when it can't find the appropriate callback. Is the saving and restoration of the validation minibatch source checkpoint state required for cross validation? And if it is, is there a workaround to getting cross validation to work in a `TrainingSession` with a `UserMinibatchSource`. Thanks!

**Comment by 1ytic** [comment]
Just add this functions in your custom `UserMinibatchSource` 
```
from cntk.internal.utils import _py_dict_to_cntk_dict

    def get_checkpoint_state(self):
        return _py_dict_to_cntk_dict({})

    def restore_from_checkpoint(self, checkpoint):
        pass
```

**Comment by eldakms** [comment]
Meanwhile I will fix the error message, UserMinibatchSource should throw by default "non implemented". Thanks!

**Comment by bencherian** [comment]
Thanks for the suggestions
@1ytic Unfortunately `_py_dict_to_cntk_dict` is not in `cntk.internal.utils` in 2.0rc2, but is only in master. However, since I don't actually need to convert a dictionary I was able to just construct an empty CNTK dictionary directly.

```Python
    def get_checkpoint_state(self):
        from cntk.cntk_py import Dictionary
        return Dictionary()

    def restore_from_checkpoint(self, checkpoint):
        pass
````

**Comment by bencherian** [comment]
OK, my code now runs, but gets stuck in an actual infinite loop upon the first cross validation call...I added a logging call in my `next_minibatch` function of the `UserMinibatchSource` and I can see that upon the first call to the cross validation routine, it gets stuck and I only end up requesting validation minibatches.

In the C++ code `TestMinibatch` looks like it might always return `true` if there was data to test, which would result in looping through the validation minibatch source (or at least until the source returns empty minibatches)...is this the intended behavior? This doesn't seem suitable if we want to cross validate several times throughout the training session...(or else we have to specifically design the `MinibatchSource` to return an empty minibatch to indicate the end of a single cross validation pass...which seem unwieldly...the MinibatchSource already provides a sweepEnd parameter that could be used for this purpose.

**Comment by eldakms** [comment]
In RC2 the cross validation runs till the empty minibatch is returned. If your minibatch source does not have an option to limit the max number of samples (or sweeps) as the built in MB sources have, then you will run forever.

If you take the latest from master you should be able to specify max_count in the CrossValidationConfig, then (no matter if the MB source has the described above logic) cross validation will stop after seen max_count samples.

**Comment by bencherian** [comment]
Great! Thanks!

Just FYI, for anyone else who might be trying to use RC2 in this manner. An empty minibatch is specified by returning an empty dictionary from `next_minibatch`...(not a `MinibatchData` object with 0 samples...which causes some weird behavior instead).

**Comment by bencherian** [comment]
I'm closing this because the issue is completely fixed in 2.0rc3.

Just as a reference to any future readers, an implementation of `UserMinibatchSource` now requires you to return a **Python** dictionary from `get_checkpoint_state` not a CNTK dictionary as given in the earlier examples. The baseclass has internal methods which handle the conversion from Python dictionaries to CNTK dictionaries.
